syntax = "proto3";

import "google/protobuf/timestamp.proto";

enum CellMethod {
  CELL_METHOD_UNSPECIFIED = 0;  // Required default
  POINT = 1;                    // Instantaneous value at a point
  SUM = 2;                      // Sum over the cell
  MAXIMUM = 3;                  // Maximum value
  MAXIMUM_ABSOLUTE_VALUE = 4;   // Maximum of absolute values
  MEDIAN = 5;                   // Median value
  MID_RANGE = 6;                // Average of maximum and minimum
  MINIMUM = 7;                  // Minimum value
  MINIMUM_ABSOLUTE_VALUE = 8;   // Minimum of absolute values
  MEAN = 9;                     // Mean (average) value
  MEAN_ABSOLUTE_VALUE = 10;     // Mean of absolute values
  MEAN_OF_UPPER_DECILE = 11;    // Mean of the upper group of data values defined by the upper tenth of their distribution
  MODE = 12;                    // Mode (most common value)
  RANGE = 13;                   // Absolute difference between maximum and minimum
  ROOT_MEAN_SQUARE = 14;        // Root mean square (RMS)
  STANDARD_DEVIATION = 15;      // Standard deviation within the cell
  SUM_OF_SQUARES = 16;          // Sum of squares
  VARIANCE = 17;                // Variance within the cell
}

message DeviceRef {
  oneof target {
    google.protobuf.Empty host = 1; // if field is present => refers to the single host
    uint32 observerId = 2;
  }
}

message Parameter {
  string longName = 1;
  string unit = 2;
  DeviceRef device = 3;
  CellMethod cellMethod = 4;
  uint32 cellPeriodSeconds = 5;
  optional map<string, string> names = 6;
}

message ParameterDefinition {
  uint32 id = 1;
  string description = 2;
  repeated Parameter parameters = 3;
  optional map<string, string> namespaces = 4;
}

message Value {
  oneof kind {
    float floatValue = 1;
    double doubleValue = 2;
    int32 intValue = 3;
    uint32 unsignedIntValue = 4;
    int64 int64Value = 5;
    uint64 unsignedInt64Value = 6;
    string stringValue = 7;
    bool boolValue = 8;
    google.protobuf.Empty emptyValue = 9;
  }
}

message Observation {
  uint32 parameterDefinitionId = 1;
  google.protobuf.Timestamp time = 2;
  repeated Value values = 3;
}

enum ReferenceSurface {
  REFERENCE_SURFACE_UNSPECIFIED = 0; // Default
  MSL = 1;                          // Mean Sea Level
  GEOID = 2;                        // Geoid
  GL = 3;                           // Ground Level
  REFERENCE_ELLIPSOID = 4;          // Reference Ellipsoid
  PRESSURE_1000_HPA = 5;            // 1000 hPa Pressure Level
}

message Location {
  optional double latitude = 1;
  optional double longitude = 2;
  double heightMeter = 3;
  ReferenceSurface referenceSurface = 4; 
}

message HostDevice {
  string name = 1;
  optional Location location= 2;
  optional string url = 3;
  optional string serialNumber = 4;
  optional string firmwareVersion = 5;
}

message ObserverDevice {
  uint32 id = 1;
  string name = 2;
  optional Location location= 3;
  optional string url = 4;
  optional string serialNumber = 5;
  optional string firmwareVersion = 6;
}

message Data {
  uint32 version = 1;
  repeated Observation observations = 2;
}

message Metadata {
  uint32 version = 1;
  HostDevice host = 2;
  repeated ObserverDevice observers = 3;
  repeated ParameterDefinition parameterDefinitions = 4;
}
